File: python/builder.py
# https://sourcemaking.com/design_patterns/builder
import json

class Burger():
    def __init__(self, meat, buns, cheese, sauces):
        self.meat = meat
        self.buns = buns
        self.cheese = cheese
        self.sauces = sauces
    
    def __repr__(self):
        return str(self.__dict__)

fries = "some fries"
drink = "Coke"
toy = "Batman"

class HappyMeal():
    def add_burger(self, burger):
        self.burger = burger
        return self
    
    def add_fries(self, fries):
        self.fries = fries
        return self
    
    def add_drink(self, drink):
        self.drink = drink
        return self
    
    def add_toy(self, toy):
        self.toy = toy
        return self
    
    def __repr__(self):
        return str(self.__dict__)

happy_meal = HappyMeal()
burger = Burger(True, True, False, False)

happy_meal.add_burger(burger).add_drink(drink)

if(fries):
    happy_meal.add_fries(fries)

happy_meal.add_toy(toy)

print(happy_meal)

File: typescript/builder.ts
// https://sourcemaking.com/design_patterns/builder

class Burger {
    meat: boolean
    buns: boolean
    cheese: boolean
    sauces: boolean

    constructor() {

    }
}

type Fries = Object;
type Drink = "Coke" | "Sprite" | "Water"
type Toy = "Batman" | "Superman"


class HappyMeal {
    burger: Burger; 
    fries: Fries;
    drink: Drink;
    toy: Toy;

    addBurger(burger: Burger) {
        this.burger = burger
        return this // return this to allow function chaining! Fancy!
    }

    addFries(fries: Fries) {
        this.fries = fries
        return this
    }

    addDrink(drink: Drink) {
        this.drink = drink
        return this
    }

    addToy(toy: Toy) {
        this.toy = toy
        return this
    }
}


// Lets create a happy meal
const happyMeal = new HappyMeal();

// Classic way of instantiating an object, must specify all portions
const burger: Burger = {
    meat: true,
    buns: true,
    cheese: false,
    sauces: false
}

const fries = undefined

// Builder pattern allows us to conditionally add portions to the object
happyMeal.addBurger(burger)
         .addDrink('Coke')

if(fries) happyMeal.addFries(fries)

happyMeal.addToy('Superman')